### ISSUE TITLE
Doctype: tile

### ISSUE THREAD

**Issue by oed** [ceramicnetwork/CIP#56]
```
cip: <to be assigned>
title: Doctype: tile
author: Joel Thorstensson @oed
discussions-to: <URL of the Github issue for this CIP (if this is a PR)>
status: Draft
category: Standards
type: Core
created: 2020-07-22
requires: 37
```

## Simple Summary

This CIP describes the `tile` doctype which is a very general document type that has JSON as it's content and is controlled by a [DID](https://w3c.github.io/did-core/).

## Abstract

The `tile` doctype allows DIDs to create simple JSON documents that uses [json-patch](http://jsonpatch.com/) for updating the content of the document. Each Tile is owned by one or multiple DIDs, all of which need to sign the update record. The signed records are encoded using. [dag-jose](https://github.com/ceramicnetwork/js-dag-jose) which allows the signatures to be natively encoded directly in [IPLD](ipld.io). Tiles also support adding schemas to the document which means that the json data within the content of the document can be enforced to have a specific format.

## Motivation

The `tile` doctype is a very generic doctype that gives developers flexibility in how they structure the data that they put into Ceramic. While other doctypes might have certain restrictions around what type of data is allowed to be stored and what the structure of that data should be, tiles allows any type of JSON data to be stored and updated. Tiles can be owned and controlled by any DID that implements the [DID Provider](https://github.com/ceramicnetwork/CIP/issues/4) specification. This provides the entire Decentralized Identity ecosystem with an interoperable verifiable document network.

## Specification

Below the record formats and state transitions of the `tile` doctype are specified. Together they should include all information needed to implement this doctype in Ceramic.

### Record formats

The **Genesis record** has the following required properties:

* `doctype` - MUST equal `tile`

* `data` - the initial JSON content of the document
* `owners` - the owners array in the header needs to contain only DID strings

In addition to the required properties above the genesis record may inlcude `schema`, `tag`, and `unique`. The `unique` property should include a random number that ensures that there is no other document that is exactly the same.

Finally the genesis record is signed and encoded using [dag-jose](https://github.com/ceramicnetwork/js-dag-jose).

**Signed records** of a `tile` document are encoded using [dag-jose](https://github.com/ceramicnetwork/js-dag-jose). Other than that they conform to the standard record format. The content of the `data` property should be a [json-patch](http://jsonpatch.com/) object which contains the difference from the previous document content state.

### State transitions

Below the state transition logic for the three different supported records is described.

#### Applying the Genesis Record

When the genesis record is applied a `DocState` object is created that looks like this:

```js
{
  doctype: 'tile',
  metadata: {
    owners: genesisRecord.owners,
    schema: genesisRecord.schema,
    tags: genesisRecord.tags
  },
  content: genesisRecord.data,
  signature: SignatureStatus.SIGNED,
  anchorStatus: AnchorStatus.NOT_REQUESTED,
  log: [new CID(genesisRecord)]
}
```

In addition to this the following should be verified:

* `owners` is an array of DID strings
* If `schema` is defined it should be Ceramic DocId string
* If `tags` is defined it should be an array of strings
* `content` should be JSON
* If `schema` is defined the data in the `content` should be valid according to the schema
* The **dag-jose** signature should be valid and signed by the DID(s) in the `owners` array

#### Applying a Signed Record

When a signed record is applied the `DocState` object should be modified in the following way:

```js
state.nextMetadata = {
  schema: signedRecord.schema,
  tags: signedRecord.tags
}
state.nextContent = applyJSONPatch(state.content, signedRecord.data)
state.signature = SignatureStatus.SIGNED
state.anchorStatus = AnchorStatus.NOT_REQUESTED
state.log.push(new CID(signedRecord))

```

Where `applyJSONPatch` is a function that applies a json-patch and returns a new json object. In addition the following should be validated:

* If `schema` is defined it should be Ceramic DocId string
* If `tags` is defined it should be an array of strings
* If `schema` is defined the data in the `content` should be valid according to the schema
* The **dag-jose** signature should be valid and signed by the DID(s) in the `owners` array in the `metadata` property (not the `nextMetadata` property of the state)

#### Applying an Anchor Record

When an anchor record is applied the `DocState` object should be modified in the following way:

```js
state.anchorStatus = AnchorStatus.ANCHORED
state.anchorProof = anchorRecord.proof
state.content = state.nextContent
delete state.nextContent
state.metadata = Object.assign(state.metadata, state.nextMetadata)
delete state.nextMetadata
state.log.push(new CID(anchorRecord))
```

No additional validation needs to happen at this point in the state transition function.

## Rationale

The `tile` specification provides a fairly straight forward way of implemting a general JSON document in Ceramic. By allowing a Schema to be defined the content of the document can be constrained. This gives developers both flexibility and control. There should not really be anything in this specification that is controversial.

## Implementation

A reference implementation of the `tile` doctype is provided in [js-ceramic](https://github.com/ceramicnetwork/js-ceramic).

## Security Considerations

Because of the double spend problem a change of `owner` is not allowed in the `tile` doctype. If ownership change was allowed a previous owner of the document could create two different document updates and only publish the latest one. Then only at a later time reclaim ownership of the specific tile document. In future versions of the ceramic protocol it will be possible to specify an NFT as an owner of a document which means that the ownership transfer can happen on-chain and thus sidestep this problem.

## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by simonovic86** [comment]
Looks great! Suggestions:

- `The unique property should include a random number that ensures that there is no other document that is exactly the same.` 
Do we need to restrict it to numbers only?
- Should we consider introducing a property `next` with embedded `content` and `metadata` instead of having separate properties `nextContent` and `nextMetadata`?

**Comment by michaelsena** [comment]
A few questions/comments/thoughts:
- Should tile be capitalized or no? I think it can be capitalized everywhere except when written as code (`tile`) and referenced by name in the header of a new document.
- This isn't a Core CIP, it's just an RFC since specific doctypes are not part of the core protocol.
- Same as Janko re: content and metadata.

**Comment by oed** [comment]
Updated to reflect your comments @simonovic86. Also updated https://github.com/ceramicnetwork/CIP/issues/37

Thanks for making all of those edits @michaelsena 🙏

**Comment by michaelsena** [comment]
@oed Should in the implementations section should we link directly to the ceramic-doctype-tile package in the js-ceramic repo?

**Comment by oed** [comment]
Good point @michaelsena I'm hesitant to link that though since we talked about changing the name of the package so the link might break. Adding some additional context instead.
